sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -210,8 +206,15 @@ object Cast {
     case _ => false  // overflow
   }
 
+  /**
+   * Returns `true` iff it should change the nullability of this type in map type,
+   * array type, expressions such as cast, etc.
+   *
+   * Note that you should take the nullability context into account.
+   * For example, it should not force to nullable type when null type in array
+   * type is non-nullable which means an empty array of null type.
+   */
   def forceNullable(from: DataType, to: DataType): Boolean = (from, to) match {
-    case (NullType, _) => true
```

Comment:
We can do this now?
```
case (NullType, _) => false // empty array or map case
```

Reply:
Hm, I think it's just better to whitelist as are rather than blacklist. `case (_, StringType) => false` could be removed I guess but let me leave it out of this PR.

Reply:
We have `case (_, DateType) => true`, so the problem still exists for `concat(array(), array(date'xxx'))`?

Reply:
I think so. I am not sure if it's right to call it problem though. We have `(_, CalendarIntervalType) => true` too

Reply:
so adding a `case (NullType, _) => false` should be good to go?

I feel it's weird if `concat(array(), array(1))` is allowed but `concat(array(), array(date'xxx'))` is not.

Reply:
Hm .. okay

Reply:
nit: I think its better to add more tests other than integer types (e.g., `null types -> date/timestamp types` discussed above) in https://github.com/apache/spark/pull/27991/files#diff-0ced8bd3a8e8459e7f66333b0d936771R1169